PXC-3470: Merge codership/galera 25.3.31 -> percona/galera 3.x

venkatesh-prasad-v opened the pull request:
https://jira.percona.com/browse/PXC-3470
    
***
Merge conflicts resolved.
    
***
Updated GALERA_VER to 3.44.
    
***
Updated GALERA-REVISION to reflect the upstream refresh revision.
    
***
Converted some percona specific unit tests to use `ck_assert*`
libcheck macro instead of the deprecated `fail_if`.
    
***
Reverted our fix for PXC-3424
"PXC-3424: Fix error handling when donor is not able to serve SST"
(https://jira.percona.com/browse/PXC-3424)
(commit 1036b2fd)
in favour of the upstream fix for the bug codership/galera#284
"Fix SST donate callback return value check"
(commit af7cd63d)

Testing Done
---
https://pxc.cd.percona.com/view/PXC%205.6/job/pxc-5.6-param/81/testReport/

Note to Reviewers
---
1. The test case `galera.galera_defaults` is failing because the [recent changes](https://galeracluster.com/2020/10/galera-cluster-for-mysql-5-6-49-5-7-31-and-8-0-21-released). This has been addressed in https://github.com/percona/percona-xtradb-cluster/pull/1309.
```
galera.galera_defaults                   [ fail ]
        Test ended at 2020-11-16 17:12:28

CURRENT_TEST: galera.galera_defaults
--- /tmp/results/PXC/mysql-test/suite/galera/r/galera_defaults.result	2020-11-16 19:21:14.000000000 +0300
+++ /tmp/results/PXC/mysql-test/var/log/galera_defaults.reject	2020-11-16 20:12:27.826260413 +0300
@@ -53,7 +53,7 @@
 WHERE VARIABLE_NAME LIKE 'wsrep_%'
 AND VARIABLE_NAME != 'wsrep_debug_sync_waiters';
 COUNT(*)
-62
+65
 SELECT VARIABLE_NAME FROM INFORMATION_SCHEMA.GLOBAL_STATUS
 WHERE VARIABLE_NAME LIKE 'wsrep_%'
 AND VARIABLE_NAME != 'wsrep_debug_sync_waiters'
@@ -81,12 +81,15 @@
 WSREP_EVS_EVICT_LIST
 WSREP_EVS_REPL_LATENCY
 WSREP_EVS_STATE
+WSREP_FLOW_CONTROL_ACTIVE
 WSREP_FLOW_CONTROL_PAUSED
 WSREP_FLOW_CONTROL_PAUSED_NS
 WSREP_FLOW_CONTROL_RECV
+WSREP_FLOW_CONTROL_REQUESTED
 WSREP_FLOW_CONTROL_SENT
 WSREP_GCACHE_POOL_SIZE
 WSREP_GCOMM_UUID
+WSREP_GMCAST_SEGMENT
 WSREP_INCOMING_ADDRESSES
 WSREP_LAST_COMMITTED
 WSREP_LOCAL_BF_ABORTS

mysqltest: Result content mismatch
```

ayurchen committed:
1. Implement index map for continuous integer sequences based on
    std::deque to use approx. 8 times less memory than std::map.
    (including unit tests and a benchmark program)
 2. Port GCache to use such map for indexing ordered buffers
    instead of std::map.

Refs codership/galera-bugs#458

temeo committed:
codership/galera#572 Use 50msec resolution in EVS unit tests.

It appeared that 1msec resolution used originally caused too tight
loop on some builders, which caused some test cases to time out.

Changed the resolution to 50msec, which seems to be good enough
resolution.

ayurchen committed:
Respect seqno_locked when attempting to discard buffers plus cleanup
and refactoring of the related code.

Refs codership/galera-bugs#490

ayurchen committed:
Rename 'random' deqmap unit test to fix compilation on FreeBSD
Refs codership/galera-bugs#494

ayurchen committed:
Convert 'Trying to continue unpaused monitor' to debug level message
Refs codership/galera-bugs#498

ayurchen committed:
Workaround for Xenial GCC compilation bug
Refs codership/galera-bugs#499

ayurchen committed:
Fix deqmap unit test compilation on funny compilers/platforms
Increase test timeouts for some unit tests

ayurchen committed:
Use `signed char` instead of `char` where it is used as an integer type.
Fixes gu::DeqMap unit tests issues on PPC and AARCH platforms.

ayurchen committed:
Additional signed char fxies that were omitted from the previous commit.

ayurchen committed:
Refs codership/galera-bugs#499
Extended the fix toc over all GCC 5 variants and below

ayurchen committed:
- write own slicing-by-8 implementation
 - write own x86 hardware accelerated implementation
 - gu_crc32c unit now is completely free of 3rd party code.
 - make internal and 3rd party implementaitons coexist
 - add internal implementations to unit tests
 - add benchmark program for CRC32C implementations

Refs codership/galera-bugs#511

ayurchen committed:
Split platform-specific code into a separate unit for future extendability
Improve CRC32C hardware support configuration and control

Refs codership/galera-bugs#511

temeo committed:
Made ASAN build pass with default unit tests enabled.

Fixed memory leaks in unit tests. Now ASAN build passes with
unit tests which are run by default.

Notable fixes:
- Backported off-by-one fix in gcs_core_send() from 4.x.
- Suppressed delayed_list_message_ leak in gcomm::evs::Node.

Unit test suppressions with ASAN build:
- test_trac_726 which is for ancient trx version.
- gcs_core suite tests which appeared to be hard to fix.

Enabled gcs_group_configuration which was previously ignored.

temeo committed:
Disable gcs_group_configuration unit test.

The test is not run completely with debug build, and fails with
assertion with release build.

temeo committed:
Fix gcs_defrag_test memleaks.

ayurchen committed:
Hardware CRC32C implementation on ARM64
Refs codership/galera-bugs#511

mend-bolt-for-github[bot] committed:
Add .whitesource configuration file

ayurchen committed:
Fix bus errors (signal 10) due to unaligned access on QEMU sparc64 platform.
Move main MMH3 implementation into .c unit for cleaner code and maintenance.

ayurchen committed:
Remove CRC32C implementation from www.evanjones.ca

Refs codership/galera-bugs#511

ayurchen committed:
Fix ASAN-detected issues with gu::DeqMap unit test.

Refs codership/galera-bugs#518

temeo committed:
Clean up remaining references to evanjones.

ayurchen committed:
Convert unit tests to use new libcheck macros instead of deprecated ones.

Refs codership/galera-bugs#531

temeo committed:
Bump version number to 3.31.

ayurchen committed:
Fix and re-enable disabled GCS group unit tests.

ayurchen committed:
Updated build scripts to handle new CMake and OpenSSL

ayurchen committed:
Add two status variables:
 - `fc_active` to indicate whether flow control is currently active
    (paused replication)
 - `fc_requested` to indicate if this node has requested replication pause

Refs codership/galera-bugs#536

ayurchen committed:
Add gmcast segment information to status variables
Remove dead stats code

Refs codership/galera-bugs#537

temeo committed:
codership/galera#577 Unit test macro usage, GCC 10 and clang fixes.

- Disabled -pedantic to get rid of `ISO C++98 does not support the 'z'
  gnu_printf length modifier [-Werror=format=]' error`.
- Added missing includes.
- Fixed ck_assert macro string format specifiers.
- Fixed bad uses of ck_assert_msg().
- Fixed non-literal string usage in ck_abort_msg() macros, reported
  by clang.
- Fixed a bug `sizeof(id - 1)` in GCS unit tests.
- Convenience header gu_inttypes.hpp to work around compiler differences.

Commit:
Fix for `gcs_destroy` to free  memory

Handle correct `gcs_sm_enter` response, if zero is returned we should not return immediately.

temeo committed:
codership/galera#284 Fix SST donate callback return value check.

Commit:
PXC-3470: Merge codership/galera 25.3.31 -> percona/galera 3.x (with conflicts)

https://jira.percona.com/browse/PXC-3470

Merge tag 'release_25.3.31' into codership_25.3.31_merge

Conflicts:
	both modified:   SConstruct
	both modified:   galera/src/replicator_smm_stats.cpp
	both modified:   galera/src/replicator_str.cpp
	both modified:   galerautils/tests/gu_fifo_test.c
	both modified:   gcs/src/unit_tests/gcs_group_test.cpp

dutow commented on gcs/src/unit_tests/gcs_group_test.cpp:
```diff
@@ -552,111 +552,55 @@ START_TEST(test_gcs_group_find_donor)
     donor = gcs_group_find_donor(&group, sv, joiner, SARGS("home3"),
-<<<<<<< HEAD
                                  &empty_uuid, GCS_SEQNO_ILL, false);
     fail_if(donor != -EHOSTDOWN);
```
wasn't this fail_if changed to ck_assert in upstream? (and a few more below)

venkatesh-prasad-v replied:
Thanks for catching this. Will fix them in the next revision.

dutow commented on galera/src/replicator_str.cpp:
```diff
@@ -338,49 +338,6 @@ ReplicatorSMM::donate_sst(void* const         recv_ctx,
                                              &state_id, 0, 0, bypass));
-<<<<<<< HEAD
-
-    /* The fix to codership/galera#284 may break backward comatibility due to
```
I don't see codership fixing this differently in the commit history. If I missed it, please mention it in the commit message. And if I didn't, and it really isn't there, this shouldn't be removed.

venkatesh-prasad-v replied:
Marked as resolved since we decided to go revert our fix and go with upstream's fix.

Commit:
PXC-3470: Merge codership/galera 25.3.31 -> percona/galera 3.x (conflicts resolved)

https://jira.percona.com/browse/PXC-3470

***
Merge conflicts resolved.

***
Updated GALERA_VER to 3.44.

***
Updated GALERA-REVISION to reflect the upstream refresh revision.

***
Converted some percona specific unit tests to use `ck_assert*`
libcheck macro instead of the deprecated `fail_if`.

***
Reverted our fix for PXC-3424
"PXC-3424: Fix error handling when donor is not able to serve SST"
(https://jira.percona.com/browse/PXC-3424)
(commit 1036b2fd)
in favour of the upstream fix for the bug codership/galera#284
"Fix SST donate callback return value check"
(commit af7cd63d)

kamil-holubicki reviewed (approved):
LGTM,
Please run the full MTR test suite before merging this and the server part.

Commit:
PXC-3470: Merge codership/galera 25.3.31 -> percona/galera 3.x (Fix test failure)

https://jira.percona.com/browse/PXC-3470

***
The test `galera_3nodes.galera_join_with_cc_A` was resulting in node's local
state seqno to be greater than the currently processing CC seqno and thus was
failing with below error in the logs

    Local state seqno (9) is greater than group seqno (7): states
    diverged. Aborting to avoid potential data loss. Remove
    '<path>/grastate.dat' file and restart if you wish to continue.

Upstream fixed this issue as part of commit id 065e484144c5999709ae8fd19844,
but however was merged incorrectly causing the test to fail.

This patch makes PXC to have same behavior as of upstream i.e, actions having
seqno less than the local seqno are skipped and discarded.

venkatesh-prasad-v commented:
Testing Done (with https://github.com/percona/percona-xtradb-cluster/pull/1309)
---
Round 1
Galera suite: https://pxc.cd.percona.com/view/PXC%205.6/job/pxc-5.6-param/82/console
Failing tests: galera_3nodes.galera_join_with_cc_A (Fixed as part of percona/galera#199)

Full suite: https://pxc.cd.percona.com/view/PXC%205.6/job/pxc-5.6-param/83/console

Round 2
Galera and sys_vars: https://pxc.cd.percona.com/view/PXC%205.6/job/pxc-5.6-param/84/console
Failing tests: None

Full suite: https://pxc.cd.percona.com/view/PXC%205.6/job/pxc-5.6-param/85/console [In Progress]
Failing tests

Failing tests:
--- PXC-3279
rpl.rpl_stm_mixed_mts_rec_crash_safe_checksum
rpl.rpl_stm_mixed_mts_rec_crash_safe
rpl.rpl_stm_mixed_mts_rec_crash_safe_small
rpl.rpl_stm_mixed_mts_crash_safe
rpl.rpl_stm_mixed_crash_safe
rpl.rpl_crash_safe_master
rpl.rpl_row_mts_rec_crash_safe
rpl.rpl_row_mts_crash_safe
rpl.rpl_row_crash_safe
rpl.rpl_mts_relay_log_post_crash_recovery
innodb.innodb_bug59641

--- PXC-3288
main.alter_table

--- PXC-3289:
tc-heuristic-recover

--- PXC-3283:
innodb.innodb-autoinc

kennt-percona reviewed (approved):
lgtm
